newIDE/app/src/InstancesEditor/TileSetVisualizer.js

```diff
@@ -0,0 +1,640 @@
+  const classes = useStylesForTile(!!highlighted);
+  return (
+    <div
+      className={classes.tile}
+      style={{
+        left: x * size,
+        top: y * size,
+        width: size * width,
+        height: size * height,
+      }}
+      // TODO: find a way to display title on mobile.
+      title={title}
+    />
+  );
+};
+
+export type TileMapTileSelection =
+  | {|
+      kind: 'single',
+      coordinates: TileMapCoordinates,
+      flipHorizontally: boolean,
+      flipVertically: boolean,
+    |}
+  | {|
+      kind: 'multiple',
+      coordinates: TileMapCoordinates[],
+    |}
+  | {|
+      kind: 'erase',
+    |};
+
+type Props = {|
+  project: gdProject,
+  objectConfiguration: gdObjectConfiguration,
+  tileMapTileSelection: ?TileMapTileSelection,
+  onSelectTileMapTile: (?TileMapTileSelection) => void,
+  allowMultipleSelection: boolean,
+  showPaintingToolbar: boolean,
+  interactive: boolean,
+  onAtlasImageLoaded?: (
+    e: SyntheticEvent<HTMLImageElement>,
+    atlasResourceName: string
+  ) => void,
+|};
+
+const TileSetVisualizer = ({
+  project,
+  objectConfiguration,
+  tileMapTileSelection,
+  onSelectTileMapTile,
+  allowMultipleSelection,
+  showPaintingToolbar,
+  interactive,
+  onAtlasImageLoaded,
+}: Props) => {
+  const forceUpdate = useForceUpdate();
+  const atlasResourceName = objectConfiguration
+    .getProperties()
+    .get('atlasImage')
+    .getValue();
+  const [
+    shouldFlipVertically,
+    setShouldFlipVertically,
+  ] = React.useState<boolean>(false);
+  const [
+    shouldFlipHorizontally,
+    setShouldFlipHorizontally,
+  ] = React.useState<boolean>(false);
+  const [
+    lastSelectedTile,
+    setLastSelectedTile,
+  ] = React.useState<?TileMapCoordinates>(null);
+  const tileContainerRef = React.useRef<?HTMLDivElement>(null);
+  const columnCount = parseFloat(
+    objectConfiguration
+      .getProperties()
+      .get('columnCount')
+      .getValue()
+  );
+  const rowCount = parseFloat(
+    objectConfiguration
+      .getProperties()
+      .get('rowCount')
+      .getValue()
+  );
+  const [clickStartCoordinates, setClickStartCoordinates] = React.useState<?{|
+    x: number,
+    y: number,
+  |}>(null);
+  const [
+    rectangularSelectionTilePreview,
+    setRectangularSelectionTilePreview,
+  ] = React.useState<?{|
+    topLeftCoordinates: TileMapCoordinates,
+    width: number,
+    height: number,
+  |}>(null);
+
+  const [hoveredTile, setHoveredTile] = React.useState<?{
+    x: number,
+    y: number,
+  }>(null);
+
+  const imageWidth = tileContainerRef.current
+    ? parseFloat(
+        getComputedStyle(tileContainerRef.current).width.replace('px', '')
+      )
+    : 0;
+  const displayedTileSize = imageWidth ? imageWidth / columnCount : null;
+
+  React.useEffect(
+    () => {
+      forceUpdate();
+    },
+    // Force update component after first mount to make sure displayedTileSize
+    // can be computed after ref has been set.
+    [forceUpdate]
```

Comment:
If it's after first mount, then leaving this array empty makes more sense?

Reply:
I don't feel comfortable making this array empty. `forceUpdate` is not supposed to change so it works